src/dd/util.rs

```diff
@@ -0,0 +1,180 @@
+    let in_: &mut io::Read = if is_inp_std {
+        sin = io::stdin();
+        &mut sin
+    } else {
+        let inp = File::open(&i_name);
+        if inp.is_err() {
+            err(format!("cannot open {}", i_name));
+        }
+        i_f = inp.unwrap();
+        &mut i_f
+    };
+    let out: &mut io::Write = if is_out_std {
+        sout = io::stdout();
+        &mut sout
+    } else {
+        let out = File::create(&o_name);
+        if out.is_err() {
+            err(format!("cannot open {}", o_name));
+        }
+        o_f = out.unwrap();
+        &mut o_f
+    };
+
+    let mut buf = vec![0u8; cmp::min(inp_buf_size, out_buf_size) as usize];
+    let mut total = 0;
+    'read_all_blocks: for _ in 0..count {
+        let mut pos = 0;
+        let mut left = inp_buf_size;
+        let mut session = 0;
+        'read_block: loop {
+            if pos == inp_buf_size {
+                break 'read_block;
+            }
+            let mut is_read_0 = false;
+            match in_.read(&mut buf[pos..pos + left]) {
+                Ok(0) => {
+                    is_read_0 = true;
+                }
+                Ok(num) => {
+                    session += num;
+                    {
+                        let mut conf = cf.lock().unwrap();
+                        total += num;
+                        if num < inp_buf_size {
+                            conf.in_.part += 1;
+                            pos += num;
+                            left = inp_buf_size - pos;
+                        } else if num == inp_buf_size {
+                            conf.in_.full += 1;
+                        } else {
+                            err(format!("overread: {} over {}", num, inp_buf_size));
+                        }
+                    }
+                }
+                Err(_) => {
+                    // TODO: handle
+                }
+            }
+            if session == inp_buf_size || is_read_0 {
+                break 'read_block;
+            }
+        }
+        let num = cmp::min(session, out_buf_size);
+        let mut pos = 0;
+        loop {
+            if pos == num {
+                break;
+            }
+            if let Some(m) = dd_out(&mut buf, pos, num - pos, out) {
+                {
+                    let mut conf = cf.lock().unwrap();
+                    conf.written += m;
+                    if m < conf.out.buf_size {
+                        conf.out.part += 1;
+                    } else {
+                        conf.out.full += 1;
+                    }
+                }
+
+                pos += m;
+            } else {
+                err("error occured".to_string()); // FIXME: make err handling
+            }
+        }
+        if total >= inp_buf_size * count {
+            break;
+        }
+    }
+    display_progress(Arc::clone(&cf));
+}
+
+pub fn run(cf: Arc<Mutex<Config>>, _sdone: chan::Sender<()>) {
+    copy(Arc::clone(&cf));
+}
+
+pub fn err(s: String) {
```

Comment:
There's actually a macro for this in `uucore` (we should probably document this somewhere).  You can use `crash!()` or `crash_if_err!()` to do the same thing as this function, but with a given exit code and a standard error prefix (something like `dd: error:` in this case).

Reply:
Document it, please, many will benefit.

Reply:
👍 

I'll try to do so soon.